File: src/main/java/com/kspt/pms/controller/UserController.java
package com.kspt.pms.controller;

import com.kspt.pms.entity.*;
import com.kspt.pms.exception.ProjectAlreadyExistsException;
import com.kspt.pms.exception.UnknownRequestParamValueException;
import com.kspt.pms.exception.NotFoundException;
import com.kspt.pms.exception.UserAlreadyExistsException;
import com.kspt.pms.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;

/**
 * Created by kivi on 03.12.17.
 */
@RestController
public class UserController {
    @Autowired
    UserRepository userRepository;
    @Autowired
    MessageRepository messageRepository;
    @Autowired
    ProjectRepository projectRepository;
    @Autowired
    TicketRepository ticketRepository;
    @Autowired
    BugReportRepository bugReportRepository;

    @RequestMapping("rest/user/{login}/authenticate")
    public String authenticate(@PathVariable String login,
                               @RequestParam("passwd") String passwd) {
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        if (!user.getPassword().equals(passwd))
            throw new NotFoundException(login);
        return "true";
    }

    @RequestMapping("rest/user/{login}")
    public User getUser(@PathVariable String login) {
        return userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
    }

    @RequestMapping(value = "rest/user/{login}", method = RequestMethod.POST)
    public void addUser(@PathVariable String login, @RequestBody User user) {
        if (userRepository.findByLogin(login).isPresent())
            throw new UserAlreadyExistsException(login);
        userRepository.save(user);
    }

    @RequestMapping("rest/user/{login}/messages")
    public Collection<Message> getMessages(@PathVariable String login) {
        return messageRepository.findByOwnerLogin(login);
    }

    @RequestMapping("rest/user/{login}/projects")
    public Collection<Project> getProjects(@PathVariable String login,
                                           @RequestParam("type") String type) {
        Collection<Project> result;
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        switch (type) {
            case "manager": result = projectRepository.findByManagerLogin(login); break;
            case "teamlead": result = projectRepository.findByTeamLeaderLogin(login); break;
            case "dev": result = projectRepository.findByDevelopersContaining(user); break;
            case "tester": result = projectRepository.findByTestersContaining(user); break;
            default: throw new UnknownRequestParamValueException("type", type);
        }
        return result;
    }

    @RequestMapping(value = "rest/user/{login}/projects", method = RequestMethod.POST)
    public void createProject(@PathVariable String login, @RequestBody Project project) {
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        if (projectRepository.findByName(project.getName()).isPresent())
            throw new ProjectAlreadyExistsException(project.getName());
        project.setManager(user);
        projectRepository.save(project);
    }

    @RequestMapping("rest/user/{login}/managed_tickets")
    public Collection<Ticket> getManagedTickets(@PathVariable String login) {
        return ticketRepository.findByCreatorLogin(login);
    }

    @RequestMapping("rest/user/{login}/assigned_tickets")
    public Collection<Ticket> getAssignedTickets(@PathVariable String login) {
        User user = userRepository.findByLogin(login)
                .orElseThrow(() -> new NotFoundException(login));
        return ticketRepository.findByAssigneesContaining(user);
    }

    @RequestMapping("rest/user/{login}/managed_reports")
    public Collection<BugReport> getManagedReports(@PathVariable String login) {
        return bugReportRepository.findByCreatorLogin(login);
    }

    @RequestMapping("rest/user/{login}/assigned_reports")
    public Collection<BugReport> getAssignedReports(@PathVariable String login) {
        return bugReportRepository.findByDeveloperLogin(login);
    }
}

File: web/resources/js/userController.js
/**
 * Created by kivi on 13.12.17.
 */

function UserService($resource) {
    return $resource('rest/user/:login', { login: '@login' });
}

function MessageService($resource) {
    return $resource('rest/user/:login/messages', { login: '@login' });
}

function UserProjectService($resource) {
    return $resource('rest/user/:login/projects?type=:type', {login: '@login', type: '@type'});
}

function UserManReportsService($resource) {
    return $resource('rest/user/:login/managed_reports', {login: '@login'});
}

function UserAssReportsService($resource) {
    return $resource('rest/user/:login/assigned_reports', {login: '@login'});
}

function UserManTicketsService($resource) {
    return $resource('rest/user/:login/managed_tickets', {login: '@login'});
}

function UserAssTicketsService($resource) {
    return $resource('rest/user/:login/assigned_tickets', {login: '@login'});
}

function UserController($scope, $routeParams,
                        UserService,
                        MessageService,
                        UserProjectService,
                        UserManReportsService,
                        UserAssReportsService,
                        UserManTicketsService,
                        UserAssTicketsService,
                        InfoShareService) {
    var url = function () {
        return {login:$routeParams.login};
    };
    var project_url = function (proj_type) {
        return {login:$routeParams.login, type: proj_type};
    };
    if (InfoShareService.getUser().login) {
        this.authlogin = InfoShareService.getUser().login;
        this.isMe = false;
    } else {
        this.authlogin = $routeParams.login;
        this.isMe = true;
    }

    this.instance = UserService.get(url());
    this.messages = MessageService.query(url());
    this.managedProjects = UserProjectService.query(project_url("manager"));
    this.leadedProjects = UserProjectService.query(project_url("teamlead"));
    this.developedProjects = UserProjectService.query(project_url("dev"));
    this.testedProjects = UserProjectService.query(project_url("tester"));

    this.managedReports = UserManReportsService.query(url());
    this.assignedReports = UserAssReportsService.query(url());
    this.managedTickets = UserManTicketsService.query(url());
    this.assignedTickets = UserAssTicketsService.query(url());

    this.updateProjects = function () {
        this.managedProjects = UserProjectService.query(project_url("manager"));
    };

    this.createProject = function () {
        if ($scope.projectName) {
            var project = new UserProjectService(project_url("manager"));
            project.name = $scope.projectName;
            project.$save(url(), function () {
                $scope.projectName = "";
                this.updateProjects();
            }.bind(this), function (error) {
                alert(error.data.message);
            });
        } else {
            alert("Enter the name of the project");
        }
    }
}

angular
    .module('app')
    .factory('UserService', UserService)
    .factory('MessageService', MessageService)
    .factory('UserProjectService', UserProjectService)
    .factory('UserManReportsService', UserManReportsService)
    .factory('UserAssReportsService', UserAssReportsService)
    .factory('UserManTicketsService', UserManTicketsService)
    .factory('UserAssTicketsService', UserAssTicketsService)
    .controller('UserController', UserController);